### ISSUE TITLE
Error in header shown in source file

### ISSUE THREAD

**Issue by steronydh** [hebaishi/linter-gcc#39]
I've got a CPP file which includes an H file. linter-gcc cannot find the header included in the H file at line 3. In Atom, it shows an error against line 3 of the CPP file against a *different* #include:

Header.h file:

```
#pragma once

#include <missing_header.h>  <<<<<< No error shown here
...
```

CPP file:

```
#include "Header.h"

#include <AnotherHeader.h> <<<< Error shown here
...
```

**Comment by hebaishi** [comment]
I have just done a small test and I am getting the same error message twice - once for the cpp file, and once for the h file.

Are you sure you are viewing all errors in the project? You can see all errors by clicking on the Project button at the bottom of the atom window. This should show you the errors in both files.

**Comment by steronydh** [comment]
If I click on the project button, I still only get the one message, but it also displays the CPP filename against the error.

**Comment by hebaishi** [comment]
So you get the same message occuring twice. Is that what you are saying?

**Comment by steronydh** [comment]
No, if I've got File selected, I get one error; if I've got Project selected, I get the same one error, but the error shows the filename as well.

**Comment by hebaishi** [comment]
Could you send me a screenshot of your atom window? Do you have both files open in atom? They both need to be open for you to see the errors from both files.

**Comment by hebaishi** [comment]
It is actually perfectly normal for an error in a header to be shown in the source file. This is because of how the C++ compiler works. The preprocessor is invoked before the compiler, so essentially all ```#include``` statements are expanded (basically the code from the included file is pasted into the current file). Then the compiler is invoked. This is why it __looks__ like the error is in the CPP file.

What I am trying to establish is whether you have both files open or only one. linter-gcc will only show errors for open files. You can find more information about C++ compilation [here](http://faculty.cs.niu.edu/~mcmahon/CS241/Notes/compile.html).

**Comment by steronydh** [comment]
I know how the C++ preprocessor works. :)  I have both header and source file open, and there is no error shown in the header file.

**Comment by hebaishi** [comment]
Ok. Could you show me the header and source files you are working with? Or maybe create a small code snippet that recreates the problem so I can play with it?

**Comment by steronydh** [comment]
I'll try and do that over the weekend. In (at least some) other compilers, if there is a problem in a header file, it shows you the error at the right line in the right file. It's odd that GCC can't do the same! :)

**Comment by MrPlatwnas** [comment]
I'm getting this error in all of my header files: "cc1: error: output filename specified twice".
When I compile the program I get no errors. It only happens inside atom when using linter-gcc.

**Comment by hebaishi** [comment]
Could you turn on debugging, open the console, copy the linter-gcc command and paste it here?

**Comment by MrPlatwnas** [comment]
<linter-message-line>cc1: error: output filename specified twice</linter-message-line>
My linter-gcc version is 0.6.8

**Comment by hebaishi** [comment]
You need to open the debugging console by pressing ```Ctrl+Shift+I```, turn on the debugging options in the linter-gcc configuration page, and copy the linter-gcc command that appears in the console window.

**Comment by MrPlatwnas** [comment]
[-1458397310067.txt](https://github.com/hebaishi/linter-gcc/files/180842/-1458397310067.txt)
Here are the log files.

**Comment by hebaishi** [comment]
Could you try pasting the command in the above file into a terminal and running it? It looks fine to me.
The command is:
```
/usr/bin/gcc -Wall -fsyntax-only -c -fmax-errors=0 /home/platwnas/Desktop/Projects/Algorithms/Recursion/tower_of_hanoi/stack.h
```

**Comment by MrPlatwnas** [comment]
$ /usr/bin/gcc -Wall -fsyntax-only -c -fmax-errors=0 /home/platwnas/Desktop/Projects/Algorithms/Recursion/tower_of_hanoi/stack.h
cc1: error: output filename specified twice

**Comment by hebaishi** [comment]
This is very strange. It looks like it's a gcc bug. Could you report the output of ```gcc --version``` and perhaps try updating GCC? Also, which platform are you on? ubuntu?

**Comment by MrPlatwnas** [comment]
$ gcc --version
gcc (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

I'm using ubuntu 15.10

GCC is up-to-date

**Comment by hebaishi** [comment]
Looks like there's something wrong with your version of GCC. Could you try installing clang and running the above command with clang instead, so substituting `clang` for `/usr/bin/gcc`?

**Comment by hebaishi** [comment]
@MrPlatwnas  unfortunately, it looks like the problem you are having is with the compiler itself. Since linter-gcc is simply an interface to GCC, it will simply display whatever GCC reports. If you disagree with me, and think linter-gcc is doing something wrong, please let me know.

@steronydh could you please do the same - paste the linter-gcc command into a terminal and see if you get the same output. If you do, the issue is with GCC itself, and not linter-gcc.

**Comment by MrPlatwnas** [comment]
I will give it a try with clang and I will report the outcome. No problem. It should be the compiler's problem, indeed.

**Comment by Arignir** [comment]
I have the same problem (headers errors in c file) quite often too

**Comment by steronydh** [comment]
I'm using platformio, and seeing:

avr-g++.exe (AVR_8_bit_GNU_Toolchain_3.4.5_1522) 4.8.1

**Comment by hebaishi** [comment]
@steronydh @Arignir do you get the same issue when your run the linter-gcc command in the terminal?

**Comment by Arignir** [comment]
Yes, but the message indicates that the error is from my header file, and not from my c file.

**Comment by steronydh** [comment]
Sorry, what linter-gcc command?

**Comment by hebaishi** [comment]
You need to do the following:
1. Turn on debugging messages in the linter-gcc configuration page
2. Press ```Ctrl+Shift+I``` to open the developer tools window
3. Click on console
4. Type in your source file to trigger linter-gcc, and the linter-gcc command will appear in the console window.

**Comment by hebaishi** [comment]
To determine the cause of this issue, I have done the following:
1. I have created two files ```source.cpp``` and ```header.hpp```. The source file contains no errors, but the header file does (a missing semicolon after a variable declaration). The source file includes the header.
2. Below is the output of running ```g++``` on both files:

```bash
/tmp/cpp_play$ cat source.cpp 
#include <iostream>
#include "header.hpp"

int main(int argc, char* argv[])
{
    return 0;
}

/tmp/cpp_play$ cat header.hpp 
#include <iostream>

int i

/tmp/cpp_play$ /usr/bin/g++ -std=c++11 -c -Wall -o /dev/null -fmax-errors=0 -I/tmp/cpp_play/include -I/tmp/cpp_play /tmp/cpp_play/source.cpp 
/tmp/cpp_play/source.cpp:4:1: error: expected initializer before ‘int’
 int main(int argc, char* argv[])
 ^

/tmp/cpp_play$ /usr/bin/g++ -std=c++11 -c -Wall -o /dev/null -fmax-errors=0 -I/tmp/cpp_play/include -I/tmp/cpp_play /tmp/cpp_play/header.hpp 
/tmp/cpp_play/header.hpp:3:5: error: expected initializer at end of input
 int i
     ^
```

You can see that when you lint the source file, the error **appears** to occur in the source file, and the same for the header. This makes perfect sense, because the C++ preprocessor does the ```#include``` substitutions before compilation. Therefore, the source file that is compiled actually has the entire contents of the header file, and this is why g++ reports an error in the source file.

My advice to you is to ensure that your headers are all error-free before working on your source files. You can easily test this by ```#include```ing your header files one by one into a empty source file to determine if one of them has an error.

**Comment by steronydh** [comment]
I understand and appreciate what is happening with the preprocessor. I've got .gcc-flags.json pointing linter-gcc at the same copy of gcc as I'm using to compile the project, and the project compiles with no errors, so I'm at a loss as to why linter-gcc should be showing errors at all!

**Comment by hebaishi** [comment]
How are you compiling your project? What build system are you using? And what is the error that you are seeing in linter-gcc that doesn't appear when you compile?

**Comment by hebaishi** [comment]
Also, you need to follow the steps I outlined describing how to find the command run by linter-gcc.

**Comment by Arignir** [comment]
@hebaishi I do understand how include works, but it could be cool if errors from the .h file could not be displayd on a .c / .cpp file, because having a red dot with an error like "missing semicolon" on a blank space in a .cpp file is very strange and quite frustrating when you get started with linter-gcc.

It will be much better if the error is not show on a c/cpp file, or at the #include line.

(Sorry for english, i'm french ^^)

**Comment by hebaishi** [comment]
@Arignir Unfortunately this is how GCC works - the preprocessor runs before the compiler. There might be a setting to get around this, but I doubt it. In any case this is a compiler issue and has nothing to do with linter-gcc.

**Comment by steronydh** [comment]
I'm using platformio which compiles the Arduino project fine.

**Comment by Arignir** [comment]
@hebaishi When the errors pops out, the file is indicate. You can juste parse this to pop the error on the good file and not always on the current file.

**Comment by hebaishi** [comment]
@Arignir This is exactly what GCC does already. That is what I tried to explain in the long post above with the two sample files.

**Comment by hebaishi** [comment]
@steronydh I will have a play with platformio and get back to you.

**Comment by Arignir** [comment]
@hebaishi >____<

We're not arrived :D

So, when there is an error in a hpp file, gcc points out that the error is on the hpp file. BUT, linter - gcc shows it in the cpp file, not in the hpp file. Is it possible to only show errors that GCC points out to be on this file, so that errors from hpp file will not be saw in cpp file ?

Thanks in advance, and sorry for my bad english ><

**Comment by hebaishi** [comment]
@Arignir I really do not think that is possible. I could be wrong, but I seriously doubt it. Expansion of ```#include``` directives is fundamental to the way the compiler works. The directives are always expanded **BEFORE** compilation, so the compiler doesn't know where the error has come from.

**Comment by Arignir** [comment]
Okey, let's have a test:
I have helloworld.c, with content:
```
#include "helloworld.h"

int main()
{
  printf("Hello World !\n");
  return (0);
}
```

And helloworld.h, with content
```
#ifndef HELLO_WORLD_H_
# define HELLO_WORLD_H_

# include <stdio.h>

Oh shit ! An error, just here !

#endif /* !HELLO_WORLD_H_ */
```

Let's now compile with the command `gcc helloworld.c -I./` (I'm using gcc 5.3)
Here is the output : 
```
In file included from helloworld.c:1:0:
helloworld.h:6:1: error: unknown type name ‘Oh’
 Oh shit ! An error, just here !
 ^
helloworld.h:6:9: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘!’ token
 Oh shit ! An error, just here !
```

Don't you see these beautiful "helloworld.h" in the error output ? ;)

**Comment by hebaishi** [comment]
Looks like we have found an inconsistency in the way GCC reports errors. In my test case, GCC reports an error in the source file, even though it's actually in the header. In the example you have given, you are right - the error is reported as coming from the header file.

Your input has been helpful in identifying the problem, which is in the case of linting on-the-fly (where linter-gcc actually lints a temporary file), the temporary filename has to be substituted for the original file, but that has to leave any header errors alone. I will update linter-gcc asap and let you know.

**Comment by Arignir** [comment]
Thanks man ! :)
Because i'm using linter-gcc 10h per day, i have tons of ideas to improve it ! Maybe i'll fork / pull request it one day :D

Anyway, thank you !

**Comment by hebaishi** [comment]
That's fantastic to hear. If it's easier for you, why don't you create another issue and make a list of some suggested improvements that we can discuss together? As someone who uses it all time, I would love to hear your ideas! Also, it would be great for all the users of linter-gcc (myself included) to benefit from these ideas as soon as possible.

**Comment by steronydh** [comment]
It looks like I need to compile Arduino code with gcc 5.3 to get the improved error reports! ;)

**Comment by hebaishi** [comment]
I have made the necessary change. Please update to linter-gcc v0.6.9. Note that a restart of Atom is necessary after updating linter-gcc.

**Comment by hebaishi** [comment]
@steronydh @Arignir everything working normally guys?

**Comment by Arignir** [comment]
@hebaishi Didn't had the time to code a lot since the update, so i don't know that much :s

**Comment by steronydh** [comment]
Looking good! Thanks, @hebaishi. Errors in header files are now showing at the exact line in the header file. :) Great fix!
